mm2src/coins/nft/storage/sql_storage.rs

```diff
@@ -63,7 +64,8 @@ fn create_tx_history_table_sql(chain: &Chain) -> MmResult<String, SqlError> {
     collection_name TEXT,
     image_url TEXT,
     token_name TEXT,
-    details_json TEXT
+    details_json TEXT,
+    PRIMARY KEY (transaction_hash, log_index, chain)
```

Comment:
**note**: I don't know about the frequency of write operations this table will get, but this PK design will definetly slow down these operations a lot specially when there are too many rows.

Reply:
I can remove this primary key all together, we already avoid writing duplicate transfers that have the same chain/transaction_hash/log_index by using the last block number when requesting new transfers that will be saved. It will be difficult though to detect problems/bugs like the one that this PR fixes if I did that.

Reply:
> we already avoid writing duplicate transfers that have the same chain/transaction_hash/log_index by using the last block number when requesting new transfers that will be saved.

If it's done with the thread-safe way, seems fine yeah.

> It will be difficult though to detect problems/bugs like the one that this PR fixes if I did that.

[The error](https://github.com/KomodoPlatform/komodo-wallet/issues/1239#issuecomment-1658128185) seems quite clear if I am not missing something.

I don't have hard feelings to remove this PK, just wanted to leave a note on it. I leave it up to you :)

Reply:
> https://github.com/KomodoPlatform/komodo-wallet/issues/1239#issuecomment-1658128185 seems quite clear if I am not missing something.

It was a db error due to the unique constraint on transaction hash, if moralis duplicates data for some reason or there are problems with our implementation that we are trying to save duplicate data it won't be caught without some kind of constraint. Adding a constraint on these values instead of a primary key will also slow down operations afaik.

> I don't have hard feelings to remove this PK, just wanted to leave a note on it. I leave it up to you :)

Let's leave it as is then and see if there are performance problems in the future :)

Reply:
> Adding a constraint on these values instead of a primary key will also slow down operations afaik.

The idea is not merging multiple values as PK/constrait because on each write each row in the table will need to check all these fields. The best way to deal with this would be creating another field(maybe `id`?) and inserting the values with pre-calculated `id`(which can be combination of the values used as PK atm) from mm2.

> Let's leave it as is then and see if there are performance problems in the future :)

Totally fine :+1:

Reply:
Aren't there a different transfer history table for each chain, why do we need the chain as part of the primary key?

Reply:
You are correct, I don't know how I missed that. I guess it's because chain is a column in the table which got me confused. Thanks for this important note, will fix it.

Reply:
Is the column necessary?

Reply:
Done

Reply:
> Is the column necessary?

It's not, but I guess it was added to make this https://github.com/KomodoPlatform/komodo-defi-framework/blob/be413ca296b1330172023d91680c8c74b0d1c1c6/mm2src/coins/nft/storage/sql_storage.rs#L201-L209 deserialize to the structs straight away. I can change the function to `transfer_history_from_row_and_chain` and insert the chain value into the json object before returning it or just leave it as is. What do you think is better?

Reply:
I think rather than keeping a field all with the same values, inserting the value while deserializing better.

Reply:
Inserting the value after deserialization and removing it before serialization would complicate the code a lot. Additionally, adding the `chain` field after fetching the row/s requires parsing the JSON string into a `serde_json::Value`, modifying it, and then deserializing it into an `Nft`/`NftTransferHistory` structs. This process likely has a greater impact on performance than simply fetching the chain value directly from the database and deserializing the entire row in one operation. Ideally, the `chain` field should not have been included in the `Nft`/`NftTransferHistory` structs from the begining, but it cannot be removed now as it is part of the responses used by GUIs.

Reply:
I can of course change these structs
https://github.com/KomodoPlatform/komodo-defi-framework/blob/717e376447f9bcb76d972b2211ad22c83e2c287a/mm2src/coins/nft/nft_structs.rs#L218-L227
https://github.com/KomodoPlatform/komodo-defi-framework/blob/717e376447f9bcb76d972b2211ad22c83e2c287a/mm2src/coins/nft/nft_structs.rs#L392-L405
to make all fields other than chain part of a new struct that is used for DB operations and use `#[serde(flatten)]` for this new struct, but the `Nft`/`NftTransferHistory` will get very complicated too. I think I would rather keep the `chain` field :)

Reply:
Okay I thought it would be simple. In that case, it can stay as it is.